### ISSUE TITLE
Model convergence differences between cpu and mps

### ISSUE THREAD

**Issue by xsatvik** [karpathy/build-nanogpt#92]
I was going through the part where Andrej was at the step where we ran a sample initial optimization run on a single batch.
When i used device as cpu, the model loss converged to  0.003, but when i used mps it converged into 8 or 9. 
The initialization loss was around 11.
I wasn't able to find out why such difference is emerging. Did this occur to anyone else, if anyone found the reason behind this let me know

I was also using torch.mps.synchronize() after each optimization step

The code untill the step is https://github.com/karpathy/build-nanogpt/blob/7822fce8ebf46ba3907ee8f7cf0356f8ec1c63fa/train_gpt2.py

**Comment by navodit17** [comment]
This happened with me too. It overfits single batch when device is 'cpu' but not on 'mps'.

Even tried running an exact copy-pasted version of train_gpt2.py to check for potential bugs in my script but got similar results.

Apple M2 Pro | MacOS: 15.4 | PyTorch: 2.6.0

**Comment by xsatvik** [comment]
don't know why synchronizing function isn't fixing it
